File: ui2/frontend/src/screens/configurationProjection.ts
/**
 * Configuration projection: the sanitized configuration a device reported,
 * turned into the operator's tables (section · setting · current value ·
 * origin · context) and the "operator snapshot" tiles -- the shape the
 * Product Owner asked for on 2026-09-22, after the old product's
 * Configuration screen.
 *
 * Pure functions over text the service already serves
 * (GET /devices/{id}/configuration/text: Check Point `set` lines with
 * secret-bearing lines withheld; Palo Alto sanitized XML with secret leaves
 * redacted). Nothing here reaches a device; nothing here invents a value --
 * a line that no rule understands lands in "Other", verbatim.
 */

export type Origin = "LOCAL" | "MEMBER" | "EFFECTIVE" | "OVERRIDE" | "PAN";

export interface SettingRow {
  readonly section: string;
  readonly setting: string;
  readonly value: string;
  readonly origin: Origin;
  readonly context: string | null;
  /** Stable key for joining two members' rows: section + setting (+ ordinal for repeated settings). */
  readonly key: string;
}

export interface Section {
  readonly label: string;
  readonly rows: readonly SettingRow[];
}

export interface SnapshotTile {
  readonly group: string;
  readonly label: string;
  readonly value: string;
  readonly origin: Origin;
}

export interface Projection {
  readonly vendor: "check_point" | "palo_alto";
  readonly sourcePlane: string;
  readonly settingCount: number;
  readonly withheldCount: number;
  readonly sections: readonly Section[];
  readonly snapshot: readonly SnapshotTile[];
}

const SECTION_ORDER = [
  "System", "DNS", "NTP", "Management", "Management Services", "Password Policy", "Login Banner", "Logging",
  "High Availability", "Interfaces", "Routing", "SNMP", "AAA", "Users", "Telemetry", "Network Configuration",
  "Settings", "Other Gaia Configuration", "Other",
];

const ACRONYMS: Record<string, string> = {
  dns: "DNS", ntp: "NTP", ssh: "SSH", ssl: "SSL", snmp: "SNMP", aaa: "AAA", arp: "ARP", ip: "IP", ipv4: "IPv4", ipv6: "IPv6",
  vs: "VS", vsx: "VSX", ha: "HA", lldp: "LLDP", rip: "RIP", ospf: "OSPF", bgp: "BGP", mtu: "MTU", lcd: "LCD", url: "URL",
  tls: "TLS", nat: "NAT", vlan: "VLAN", id: "ID", mgmt: "Mgmt",
};

function titleCase(token: string): string {
  return token
    .split(/[-_]/)
    .filter(Boolean)
    .map((t) => ACRONYMS[t.toLowerCase()] ?? t.charAt(0).toUpperCase() + t.slice(1))
    .join(" ");
}

function orderSections(bySection: Map<string, SettingRow[]>): Section[] {
  const labels = [...bySection.keys()].sort((a, b) => {
    const ia = SECTION_ORDER.indexOf(a);
    const ib = SECTION_ORDER.indexOf(b);
    return (ia === -1 ? 999 : ia) - (ib === -1 ? 999 : ib) || a.localeCompare(b);
  });
  return labels.map((label) => ({ label, rows: bySection.get(label) ?? [] }));
}

function push(bySection: Map<string, SettingRow[]>, counters: Map<string, number>, row: Omit<SettingRow, "key">): void {
  const base = `${row.section}|${row.setting}`;
  const n = (counters.get(base) ?? 0) + 1;
  counters.set(base, n);
  const key = n === 1 ? base : `${base}#${n}`;
  const rows = bySection.get(row.section) ?? [];
  rows.push({ ...row, key });
  bySection.set(row.section, rows);
}

// ---------------------------------------------------------------------------------------------
// Check Point Gaia -- `set ...` lines
// ---------------------------------------------------------------------------------------------

/** One rule: the tokens after `set` that select it, the section, and how many tokens after the prefix belong to the setting name. */
interface CpRule {
  readonly prefix: readonly string[];
  readonly section: string;
  /** Tokens (after the prefix) that are part of the setting name; the rest is the value. */
  readonly nameTokens: number;
  readonly label?: (tokens: readonly string[]) => string;
  readonly origin?: Origin;
}

const CP_RULES: readonly CpRule[] = [
  { prefix: ["hostname"], section: "System", nameTokens: 0, label: () => "Hostname", origin: "MEMBER" },
  { prefix: ["domainname"], section: "System", nameTokens: 0, label: () => "Domain" },
  { prefix: ["timezone"], section: "System", nameTokens: 0, label: () => "Timezone" },
  { prefix: ["format"], section: "System", nameTokens: 1 },
  { prefix: ["dns", "primary"], section: "DNS", nameTokens: 0, label: () => "Primary DNS" },
  { prefix: ["dns", "secondary"], section: "DNS", nameTokens: 0, label: () => "Secondary DNS" },
  { prefix: ["dns", "tertiary"], section: "DNS", nameTokens: 0, label: () => "Tertiary DNS" },
  { prefix: ["dns"], section: "DNS", nameTokens: 1 },
  { prefix: ["ntp", "server", "primary"], section: "NTP", nameTokens: 0, label: () => "Primary NTP Server" },
  { prefix: ["ntp", "server", "secondary"], section: "NTP", nameTokens: 0, label: () => "Secondary NTP Server" },
  { prefix: ["ntp"], section: "NTP", nameTokens: 1 },
  { prefix: ["inactivity-timeout"], section: "Management", nameTokens: 0, label: () => "Inactivity Timeout" },
  { prefix: ["management"], section: "Management", nameTokens: 1 },
  { prefix: ["web"], section: "Management", nameTokens: 1 },
  { prefix: ["ssh"], section: "Management", nameTokens: 1 },
  { prefix: ["net-access"], section: "Management", nameTokens: 1 },
  { prefix: ["proxy"], section: "Management", nameTokens: 0, label: () => "Proxy" },
  { prefix: ["ssl"], section: "Management", nameTokens: 1 },
  { prefix: ["password-controls"], section: "Password Policy", nameTokens: 1 },
  { prefix: ["message"], section: "Login Banner", nameTokens: 1 },
  { prefix: ["syslog"], section: "Logging", nameTokens: 1 },
  { prefix: ["cluster"], section: "High Availability", nameTokens: 1 },
  { prefix: ["interface"], section: "Interfaces", nameTokens: 2, label: (t) => `Interface ${t[0]} · ${titleCase(t[1] ?? "")}`, origin: "LOCAL" },
  { prefix: ["static-route"], section: "Routing", nameTokens: 1, label: (t) => `Static Route · ${t[0]}` },
  { prefix: ["ospf"], section: "Routing", nameTokens: 1 },
  { prefix: ["rip"], section: "Routing", nameTokens: 1 },
  { prefix: ["bgp"], section: "Routing", nameTokens: 1 },
  { prefix: ["inbound-route-filter"], section: "Routing", nameTokens: 1 },
  { prefix: ["router-options"], section: "Routing", nameTokens: 1 },
  { prefix: ["routedsyslog"], section: "Routing", nameTokens: 0, label: () => "Routed Syslog" },
  { prefix: ["max-path-splits"], section: "Routing", nameTokens: 0, label: () => "Max Path Splits" },
  { prefix: ["snmp"], section: "SNMP", nameTokens: 1 },
  { prefix: ["aaa"], section: "AAA", nameTokens: 1 },
  { prefix: ["user"], section: "Users", nameTokens: 1, label: (t) => `User · ${t[0]}` },
];

const CP_MEMBER_SPECIFIC = /^(Interface .* · (IPv4 Address|IPv6 Address)|Hostname)$/;

export function projectCheckPoint(text: string): Projection {
  const bySection = new Map<string, SettingRow[]>();
  const counters = new Map<string, number>();
  let withheldMarkers = 0;
  let withheldDeclared = 0;
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line) continue;
    if (line.startsWith("#")) {
      if (/WITHHELD\]?$/i.test(line)) withheldMarkers++;
      const m = /secret-bearing-lines-withheld=(\d+)/.exec(line);
      if (m) withheldDeclared = Number(m[1]);
      continue;
    }
    if (!line.startsWith("set ")) continue;
    const tokens = line.slice(4).split(/\s+/);
    const rule = CP_RULES.find((r) => r.prefix.every((p, i) => tokens[i] === p));
    if (!rule) {
      push(bySection, counters, {
        section: "Other Gaia Configuration",
        setting: titleCase(tokens[0] ?? "") + (tokens[1] ? ` · ${titleCase(tokens[1])}` : ""),
        value: tokens.slice(Math.min(2, tokens.length)).join(" ") || "on",
        origin: "LOCAL",
        context: null,
      });
      continue;
    }
    const after = tokens.slice(rule.prefix.length);
    const nameTokens = after.slice(0, rule.nameTokens);
    const value = after.slice(rule.nameTokens).join(" ");
    const setting = rule.label
      ? rule.label(nameTokens)
      : `${titleCase(rule.prefix[rule.prefix.length - 1])}${nameTokens.length ? ` · ${nameTokens.map(titleCase).join(" · ")}` : ""}`;
    const origin: Origin = rule.origin === "MEMBER" || CP_MEMBER_SPECIFIC.test(setting) ? "MEMBER" : "LOCAL";
    push(bySection, counters, { section: rule.section, setting, value: value || "on", origin, context: null });
  }
  const withheld = Math.max(withheldMarkers, withheldDeclared);
  const sections = orderSections(bySection);
  const all = sections.flatMap((s) => s.rows);
  const pick = (section: string, setting: string): SettingRow | undefined =>
    all.find((r) => r.section === section && r.setting === setting);
  const snapshot: SnapshotTile[] = [];
  for (const [group, section, setting, label] of [
    ["SYSTEM", "System", "Hostname", "Hostname"],
    ["SYSTEM", "System", "Domain", "Domain"],
    ["SYSTEM", "System", "Timezone", "Timezone"],
    ["DNS", "DNS", "Primary DNS", "Primary DNS"],
    ["DNS", "DNS", "Secondary DNS", "Secondary DNS"],
    ["NTP", "NTP", "Primary NTP Server", "Primary NTP Server"],
    ["NTP", "NTP", "Secondary NTP Server", "Secondary NTP Server"],
    ["MANAGEMENT", "Management", "Management · Interface", "Management Interface"],
    ["HIGH AVAILABILITY", "High Availability", "Cluster · Member", "Cluster Member"],
  ] as const) {
    const row = pick(section, setting);
    if (row) snapshot.push({ group, label, value: row.value, origin: row.origin });
  }
  return { vendor: "check_point", sourcePlane: "gaia-clish-show-configuration", settingCount: all.length, withheldCount: withheld, sections, snapshot };
}

// ---------------------------------------------------------------------------------------------
// Palo Alto -- sanitized XML
// ---------------------------------------------------------------------------------------------

interface PanRule {
  readonly test: RegExp;
  readonly section: string;
}

const PAN_RULES: readonly PanRule[] = [
  { test: /\/deviceconfig\/system\/dns-setting\//, section: "DNS" },
  { test: /\/deviceconfig\/system\/ntp-servers\//, section: "NTP" },
  { test: /\/deviceconfig\/system\/(permitted-ip|service)\//, section: "Management Services" },
  { test: /\/deviceconfig\/system\/(update-schedule|panorama|snmp-setting|login-banner|motd-and-banner)/, section: "Management" },
  { test: /\/deviceconfig\/system\//, section: "System" },
  { test: /\/deviceconfig\/high-availability\//, section: "High Availability" },
  { test: /\/deviceconfig\/setting\/management\//, section: "Management" },
  { test: /\/deviceconfig\/setting\//, section: "Settings" },
  { test: /\/mgt-config\/password-complexity\//, section: "Password Policy" },
  { test: /\/mgt-config\/users\//, section: "Users" },
  { test: /\/shared\/log-settings\//, section: "Logging" },
  { test: /\/network\/interface\//, section: "Interfaces" },
  { test: /\/network\/virtual-router\//, section: "Routing" },
  { test: /\/network\//, section: "Network Configuration" },
  { test: /\/telemetry|\/statistics-service/, section: "Telemetry" },
];

const PAN_MEMBER_SPECIFIC = /\/deviceconfig\/system\/(hostname|ip-address|ipv6-address)$|\/high-availability\/interface\/[^/]+\/(ip-address|ipv6-address)$|\/high-availability\/.*peer-ip[^/]*$/;

function panLabel(path: string, names: readonly string[]): string {
  // "/config/devices/entry[localhost.localdomain]/deviceconfig/system/dns-setting/servers/primary"
  const parts = path.split("/").filter((p) => p && p !== "config" && p !== "devices" && p !== "response" && p !== "result");
  const cleaned = parts.filter((p) => !/^(deviceconfig|entry|localhost\.localdomain|vsys1)$/.test(p) && !names.includes(p));
  const nameSuffix = names.filter((n) => n !== "localhost.localdomain").map((n) => ` · ${n}`).join("");
  const label = cleaned.slice(-3).map(titleCase).join(" · ");
  if (label === "Dns Setting · Servers · Primary" || label === "DNS Setting · Servers · Primary") return "Primary DNS";
  if (label.endsWith("Servers · Secondary") && path.includes("dns-setting")) return "Secondary DNS";
  if (path.includes("ntp-servers/primary-ntp-server")) return "Primary NTP Server";
  if (path.includes("ntp-servers/secondary-ntp-server")) return "Secondary NTP Server";
  return (label || titleCase(parts[parts.length - 1] ?? "")) + nameSuffix;
}

export function projectPaloAlto(xml: string, overridePaths: readonly string[] = []): Projection {
  const bySection = new Map<string, SettingRow[]>();
  const counters = new Map<string, number>();
  let withheld = 0;
  const parser = new DOMParser();
  const doc = parser.parseFromString(xml, "application/xml");
  if (doc.getElementsByTagName("parsererror").length > 0) {
    return { vendor: "palo_alto", sourcePlane: "active", settingCount: 0, withheldCount: 0, sections: [], snapshot: [] };
  }
  const counts = { interfaces: 0, vsys: 0, virtualRouters: 0, zones: 0 };
  const walk = (el: Element, path: string, names: string[]): void => {
    const name = el.getAttribute("name");
    const here = `${path}/${el.tagName}`;
    const nextNames = name ? [...names, name] : names;
    if (/\/network\/interface\/(ethernet|aggregate-ethernet|loopback|vlan|tunnel)(\/units)?$/.test(path) && el.tagName === "entry") counts.interfaces++;
    if (/\/vsys$/.test(path) && el.tagName === "entry") counts.vsys++;
    if (/\/network\/virtual-router$/.test(path) && el.tagName === "entry") counts.virtualRouters++;
    if (/\/zone$/.test(path) && el.tagName === "entry") counts.zones++;
    const children = [...el.children];
    if (children.length === 0) {
      const value = (el.textContent ?? "").trim();
      if (/\/network\//.test(here) || /\/vsys\//.test(here) || /\/shared\//.test(here)) return; // counted, not listed
      if (value === "[REDACTED]") {
        withheld++;
        return;
      }
      const rule = PAN_RULES.find((r) => r.test.test(here));
      const section = rule?.section ?? "Other";
      // A member's own name and addresses -- system hostname / ip-address, the HA link addresses -- differ by
      // nature between cluster members (MEMBER), like a Check Point hostname or interface address.
      const memberSpecific = PAN_MEMBER_SPECIFIC.test(here);
      const origin: Origin = memberSpecific ? "MEMBER" : overridePaths.some((p) => here.includes(p)) ? "OVERRIDE" : "EFFECTIVE";
      // An element with no text is a switch or an empty container: say so, rather than a value that reads like one.
      push(bySection, counters, { section, setting: panLabel(here, nextNames), value: value || "(set, no value)", origin, context: null });
      return;
    }
    for (const child of children) walk(child, here, nextNames);
  };
  if (doc.documentElement) walk(doc.documentElement, "", []);
  for (const [setting, value] of [
    ["VSYS", counts.vsys], ["Virtual Routers", counts.virtualRouters], ["Zones", counts.zones], ["Interfaces", counts.interfaces],
  ] as const) {
    push(bySection, counters, { section: "Network Configuration", setting, value: String(value), origin: "EFFECTIVE", context: null });
  }
  const sections = orderSections(bySection);
  const all = sections.flatMap((s) => s.rows);
  const find = (section: string, pred: (s: string) => boolean) => all.find((r) => r.section === section && pred(r.setting));
  const snapshot: SnapshotTile[] = [];
  const tiles: Array<[string, string, SettingRow | undefined]> = [
    ["SYSTEM", "Hostname", find("System", (s) => s === "System · Hostname" || s.endsWith("Hostname"))],
    ["SYSTEM", "Timezone", find("System", (s) => s.endsWith("Timezone"))],
    ["DNS", "Primary DNS", find("DNS", (s) => s === "Primary DNS")],
    ["DNS", "Secondary DNS", find("DNS", (s) => s === "Secondary DNS")],
    ["NTP", "Primary NTP Server", find("NTP", (s) => s === "Primary NTP Server")],
    ["NTP", "Secondary NTP Server", find("NTP", (s) => s === "Secondary NTP Server")],
    ["HIGH AVAILABILITY", "HA Enabled", find("High Availability", (s) => s.endsWith("Enabled") && !s.includes("Link"))],
    ["HIGH AVAILABILITY", "Group ID", find("High Availability", (s) => s.endsWith("Group ID"))],
  ];
  for (const [group, label, row] of tiles) {
    if (row) snapshot.push({ group, label, value: row.value, origin: row.origin });
  }
  return { vendor: "palo_alto", sourcePlane: "effective-running", settingCount: all.length, withheldCount: withheld, sections, snapshot };
}

// ---------------------------------------------------------------------------------------------
// Cluster view -- two (or more) members side by side
// ---------------------------------------------------------------------------------------------

export interface MemberRow {
  readonly section: string;
  readonly setting: string;
  readonly key: string;
  readonly origin: Origin;
  /** member id -> value ("—" when absent on that member). */
  readonly values: Readonly<Record<string, string>>;
  /** True when two collected members hold different values and the setting is not member-specific by nature. */
  readonly diff: boolean;
  /** True when the setting is expected to differ per member (hostname, member addresses). */
  readonly memberSpecific: boolean;
}

export interface ClusterProjection {
  readonly memberIds: readonly string[];
  readonly sections: readonly { readonly label: string; readonly rows: readonly MemberRow[] }[];
  readonly diffCount: number;
  readonly settingCount: number;
}

export function projectCluster(members: ReadonlyArray<{ id: string; projection: Projection }>): ClusterProjection {
  const memberIds = members.map((m) => m.id);
  const byKey = new Map<string, { section: string; setting: string; origin: Origin; values: Record<string, string> }>();
  const order: string[] = [];
  for (const member of members) {
    for (const section of member.projection.sections) {
      for (const row of section.rows) {
        let entry = byKey.get(row.key);
        if (!entry) {
          entry = { section: row.section, setting: row.setting, origin: row.origin, values: {} };
          byKey.set(row.key, entry);
          order.push(row.key);
        }
        entry.values[member.id] = row.value;
      }
    }
  }
  const bySection = new Map<string, MemberRow[]>();
  let diffCount = 0;
  for (const key of order) {
    const e = byKey.get(key)!;
    const memberSpecific = e.origin === "MEMBER";
    const present = memberIds.map((id) => e.values[id]).filter((v): v is string => v !== undefined);
    const diff = !memberSpecific && (present.length !== memberIds.length || new Set(present).size > 1);
    if (diff) diffCount++;
    const values: Record<string, string> = {};
    for (const id of memberIds) values[id] = e.values[id] ?? "—";
    const rows = bySection.get(e.section) ?? [];
    rows.push({ section: e.section, setting: e.setting, key, origin: e.origin, values, diff, memberSpecific });
    bySection.set(e.section, rows);
  }
  const sections = orderSections(bySection as unknown as Map<string, SettingRow[]>).map((s) => ({
    label: s.label,
    rows: bySection.get(s.label) ?? [],
  }));
  return { memberIds, sections, diffCount, settingCount: order.length };
}

export function filterProjection(sections: readonly Section[], query: string): Section[] {
  const q = query.trim().toLowerCase();
  if (!q) return [...sections];
  return sections
    .map((s) => ({ label: s.label, rows: s.rows.filter((r) => r.setting.toLowerCase().includes(q) || r.value.toLowerCase().includes(q) || s.label.toLowerCase().includes(q)) }))
    .filter((s) => s.rows.length > 0);
}

File: ui2/service/src/main/java/com/securityexpert/nexus/ui2/service/overview/ConfigurationProjection.java
package com.securityexpert.nexus.ui2.service.overview;

import java.io.StringReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.regex.Pattern;

import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

/**
 * The server-side twin of {@code ui2/frontend/src/screens/configurationProjection.ts}
 * (OVERVIEW_EXCEPTION_SCREEN_CONTRACT §5.1): the same rules, the same row keys, the same MEMBER-specific
 * exemptions, so the Overview's cluster DIFF count equals what the Configuration screen shows. Only the parts
 * the DIFF count needs are ported (rows: section, setting, key, value, origin); snapshot tiles stay in the
 * browser. A shared fixture test pins the two implementations together.
 */
public final class ConfigurationProjection {

    public record Row(String section, String setting, String key, String value, boolean memberSpecific) {
    }

    public record ClusterDiff(int diffCount, int settingCount, List<String> diffSections) {
    }

    private ConfigurationProjection() {
    }

    private static final Map<String, String> ACRONYMS = Map.ofEntries(
            Map.entry("dns", "DNS"), Map.entry("ntp", "NTP"), Map.entry("ssh", "SSH"), Map.entry("ssl", "SSL"),
            Map.entry("snmp", "SNMP"), Map.entry("aaa", "AAA"), Map.entry("arp", "ARP"), Map.entry("ip", "IP"),
            Map.entry("ipv4", "IPv4"), Map.entry("ipv6", "IPv6"), Map.entry("vs", "VS"), Map.entry("vsx", "VSX"),
            Map.entry("ha", "HA"), Map.entry("lldp", "LLDP"), Map.entry("rip", "RIP"), Map.entry("ospf", "OSPF"),
            Map.entry("bgp", "BGP"), Map.entry("mtu", "MTU"), Map.entry("lcd", "LCD"), Map.entry("url", "URL"),
            Map.entry("tls", "TLS"), Map.entry("nat", "NAT"), Map.entry("vlan", "VLAN"), Map.entry("id", "ID"),
            Map.entry("mgmt", "Mgmt"));

    static String titleCase(String token) {
        List<String> out = new ArrayList<>();
        for (String t : token.split("[-_]")) {
            if (t.isEmpty()) {
                continue;
            }
            String acronym = ACRONYMS.get(t.toLowerCase(Locale.ROOT));
            out.add(acronym != null ? acronym : Character.toUpperCase(t.charAt(0)) + t.substring(1));
        }
        return String.join(" ", out);
    }

    private static final class Builder {
        final List<Row> rows = new ArrayList<>();
        final Map<String, Integer> counters = new HashMap<>();

        void push(String section, String setting, String value, boolean memberSpecific) {
            String base = section + "|" + setting;
            int n = counters.merge(base, 1, Integer::sum);
            rows.add(new Row(section, setting, n == 1 ? base : base + "#" + n, value, memberSpecific));
        }
    }

    // ---------------------------------------------------------------------------------------------- Check Point

    private record CpRule(List<String> prefix, String section, int nameTokens, Function<List<String>, String> label,
            boolean member) {
    }

    private static CpRule rule(String prefix, String section, int nameTokens) {
        return new CpRule(List.of(prefix.split(" ")), section, nameTokens, null, false);
    }

    private static CpRule rule(String prefix, String section, int nameTokens, Function<List<String>, String> label) {
        return new CpRule(List.of(prefix.split(" ")), section, nameTokens, label, false);
    }

    private static final List<CpRule> CP_RULES = List.of(
            new CpRule(List.of("hostname"), "System", 0, t -> "Hostname", true),
            rule("domainname", "System", 0, t -> "Domain"),
            rule("timezone", "System", 0, t -> "Timezone"),
            rule("format", "System", 1),
            rule("dns primary", "DNS", 0, t -> "Primary DNS"),
            rule("dns secondary", "DNS", 0, t -> "Secondary DNS"),
            rule("dns tertiary", "DNS", 0, t -> "Tertiary DNS"),
            rule("dns", "DNS", 1),
            rule("ntp server primary", "NTP", 0, t -> "Primary NTP Server"),
            rule("ntp server secondary", "NTP", 0, t -> "Secondary NTP Server"),
            rule("ntp", "NTP", 1),
            rule("inactivity-timeout", "Management", 0, t -> "Inactivity Timeout"),
            rule("management", "Management", 1),
            rule("web", "Management", 1),
            rule("ssh", "Management", 1),
            rule("net-access", "Management", 1),
            rule("proxy", "Management", 0, t -> "Proxy"),
            rule("ssl", "Management", 1),
            rule("password-controls", "Password Policy", 1),
            rule("message", "Login Banner", 1),
            rule("syslog", "Logging", 1),
            rule("cluster", "High Availability", 1),
            rule("interface", "Interfaces", 2, t -> "Interface " + (t.isEmpty() ? "" : t.get(0)) + " · "
                    + titleCase(t.size() > 1 ? t.get(1) : "")),
            rule("static-route", "Routing", 1, t -> "Static Route · " + (t.isEmpty() ? "" : t.get(0))),
            rule("ospf", "Routing", 1),
            rule("rip", "Routing", 1),
            rule("bgp", "Routing", 1),
            rule("inbound-route-filter", "Routing", 1),
            rule("router-options", "Routing", 1),
            rule("routedsyslog", "Routing", 0, t -> "Routed Syslog"),
            rule("max-path-splits", "Routing", 0, t -> "Max Path Splits"),
            rule("snmp", "SNMP", 1),
            rule("aaa", "AAA", 1),
            rule("user", "Users", 1, t -> "User · " + (t.isEmpty() ? "" : t.get(0))));

    private static final Pattern CP_MEMBER_SPECIFIC = Pattern.compile("^(Interface .* · (IPv4 Address|IPv6 Address)|Hostname)$");

    public static List<Row> checkPoint(String text) {
        Builder b = new Builder();
        for (String raw : text.split("\r?\n")) {
            String line = raw.strip();
            if (line.isEmpty() || line.startsWith("#") || !line.startsWith("set ")) {
                continue;
            }
            List<String> tokens = List.of(line.substring(4).strip().split("\\s+"));
            CpRule matched = null;
            for (CpRule r : CP_RULES) {
                boolean ok = r.prefix().size() <= tokens.size();
                for (int i = 0; ok && i < r.prefix().size(); i++) {
                    ok = r.prefix().get(i).equals(tokens.get(i));
                }
                if (ok) {
                    matched = r;
                    break;
                }
            }
            if (matched == null) {
                String setting = titleCase(tokens.get(0)) + (tokens.size() > 1 ? " · " + titleCase(tokens.get(1)) : "");
                String value = String.join(" ", tokens.subList(Math.min(2, tokens.size()), tokens.size()));
                b.push("Other Gaia Configuration", setting, value.isEmpty() ? "on" : value, false);
                continue;
            }
            List<String> after = tokens.subList(matched.prefix().size(), tokens.size());
            List<String> nameTokens = after.subList(0, Math.min(matched.nameTokens(), after.size()));
            String value = String.join(" ", after.subList(Math.min(matched.nameTokens(), after.size()), after.size()));
            String setting;
            if (matched.label() != null) {
                setting = matched.label().apply(nameTokens);
            } else {
                StringBuilder s = new StringBuilder(titleCase(matched.prefix().get(matched.prefix().size() - 1)));
                if (!nameTokens.isEmpty()) {
                    s.append(" · ").append(String.join(" · ", nameTokens.stream().map(ConfigurationProjection::titleCase).toList()));
                }
                setting = s.toString();
            }
            boolean member = matched.member() || CP_MEMBER_SPECIFIC.matcher(setting).matches();
            b.push(matched.section(), setting, value.isEmpty() ? "on" : value, member);
        }
        return b.rows;
    }

    // ---------------------------------------------------------------------------------------------- Palo Alto

    private record PanRule(Pattern test, String section) {
    }

    private static final List<PanRule> PAN_RULES = List.of(
            new PanRule(Pattern.compile("/deviceconfig/system/dns-setting/"), "DNS"),
            new PanRule(Pattern.compile("/deviceconfig/system/ntp-servers/"), "NTP"),
            new PanRule(Pattern.compile("/deviceconfig/system/(permitted-ip|service)/"), "Management Services"),
            new PanRule(Pattern.compile("/deviceconfig/system/(update-schedule|panorama|snmp-setting|login-banner|motd-and-banner)"), "Management"),
            new PanRule(Pattern.compile("/deviceconfig/system/"), "System"),
            new PanRule(Pattern.compile("/deviceconfig/high-availability/"), "High Availability"),
            new PanRule(Pattern.compile("/deviceconfig/setting/management/"), "Management"),
            new PanRule(Pattern.compile("/deviceconfig/setting/"), "Settings"),
            new PanRule(Pattern.compile("/mgt-config/password-complexity/"), "Password Policy"),
            new PanRule(Pattern.compile("/mgt-config/users/"), "Users"),
            new PanRule(Pattern.compile("/shared/log-settings/"), "Logging"),
            new PanRule(Pattern.compile("/network/interface/"), "Interfaces"),
            new PanRule(Pattern.compile("/network/virtual-router/"), "Routing"),
            new PanRule(Pattern.compile("/network/"), "Network Configuration"),
            new PanRule(Pattern.compile("/telemetry|/statistics-service"), "Telemetry"));

    private static final Pattern PAN_MEMBER_SPECIFIC = Pattern.compile(
            "/deviceconfig/system/(hostname|ip-address|ipv6-address)$|/high-availability/interface/[^/]+/(ip-address|ipv6-address)$|/high-availability/.*peer-ip[^/]*$");
    private static final Pattern PAN_DROPPED_PART = Pattern.compile("^(deviceconfig|entry|localhost\\.localdomain|vsys1)$");

    static String panLabel(String path, List<String> names) {
        List<String> parts = new ArrayList<>();
        for (String p : path.split("/")) {
            if (!p.isEmpty() && !p.equals("config") && !p.equals("devices") && !p.equals("response") && !p.equals("result")) {
                parts.add(p);
            }
        }
        List<String> cleaned = parts.stream().filter(p -> !PAN_DROPPED_PART.matcher(p).matches() && !names.contains(p)).toList();
        StringBuilder nameSuffix = new StringBuilder();
        for (String n : names) {
            if (!n.equals("localhost.localdomain")) {
                nameSuffix.append(" · ").append(n);
            }
        }
        String label = String.join(" · ", cleaned.subList(Math.max(0, cleaned.size() - 3), cleaned.size()).stream()
                .map(ConfigurationProjection::titleCase).toList());
        if (label.equals("Dns Setting · Servers · Primary") || label.equals("DNS Setting · Servers · Primary")) {
            return "Primary DNS";
        }
        if (label.endsWith("Servers · Secondary") && path.contains("dns-setting")) {
            return "Secondary DNS";
        }
        if (path.contains("ntp-servers/primary-ntp-server")) {
            return "Primary NTP Server";
        }
        if (path.contains("ntp-servers/secondary-ntp-server")) {
            return "Secondary NTP Server";
        }
        return (label.isEmpty() ? titleCase(parts.isEmpty() ? "" : parts.get(parts.size() - 1)) : label) + nameSuffix;
    }

    public static List<Row> paloAlto(String xml) {
        Document doc;
        try {
            DocumentBuilderFactory f = DocumentBuilderFactory.newInstance();
            f.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
            f.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
            f.setExpandEntityReferences(false);
            DocumentBuilder db = f.newDocumentBuilder();
            db.setErrorHandler(null);
            doc = db.parse(new InputSource(new StringReader(xml)));
        } catch (Exception e) {
            return List.of();
        }
        Builder b = new Builder();
        int[] counts = new int[4]; // interfaces, vsys, virtual routers, zones
        walk(doc.getDocumentElement(), "", List.of(), b, counts);
        b.push("Network Configuration", "VSYS", String.valueOf(counts[1]), false);
        b.push("Network Configuration", "Virtual Routers", String.valueOf(counts[2]), false);
        b.push("Network Configuration", "Zones", String.valueOf(counts[3]), false);
        b.push("Network Configuration", "Interfaces", String.valueOf(counts[0]), false);
        return b.rows;
    }

    private static final Pattern IFACE_PARENT = Pattern.compile("/network/interface/(ethernet|aggregate-ethernet|loopback|vlan|tunnel)(/units)?$");

    private static void walk(Element el, String path, List<String> names, Builder b, int[] counts) {
        String name = el.hasAttribute("name") ? el.getAttribute("name") : null;
        String tag = el.getTagName();
        String here = path + "/" + tag;
        List<String> nextNames = name != null && !name.isEmpty() ? append(names, name) : names;
        if ("entry".equals(tag)) {
            if (IFACE_PARENT.matcher(path).find()) {
                counts[0]++;
            }
            if (path.endsWith("/vsys")) {
                counts[1]++;
            }
            if (path.endsWith("/network/virtual-router")) {
                counts[2]++;
            }
            if (path.endsWith("/zone")) {
                counts[3]++;
            }
        }
        List<Element> children = new ArrayList<>();
        NodeList nodes = el.getChildNodes();
        for (int i = 0; i < nodes.getLength(); i++) {
            if (nodes.item(i).getNodeType() == Node.ELEMENT_NODE) {
                children.add((Element) nodes.item(i));
            }
        }
        if (children.isEmpty()) {
            String value = el.getTextContent() == null ? "" : el.getTextContent().strip();
            if (here.contains("/network/") || here.contains("/vsys/") || here.contains("/shared/")) {
                return;
            }
            if ("[REDACTED]".equals(value)) {
                return;
            }
            String section = "Other";
            for (PanRule r : PAN_RULES) {
                if (r.test().matcher(here).find()) {
                    section = r.section();
                    break;
                }
            }
            boolean member = PAN_MEMBER_SPECIFIC.matcher(here).find();
            b.push(section, panLabel(here, nextNames), value.isEmpty() ? "(set, no value)" : value, member);
            return;
        }
        for (Element child : children) {
            walk(child, here, nextNames, b, counts);
        }
    }

    private static List<String> append(List<String> list, String value) {
        List<String> out = new ArrayList<>(list);
        out.add(value);
        return out;
    }

    // ---------------------------------------------------------------------------------------------- Cluster

    /** {@code projectCluster}: a key present on some members only, or with differing values, is a DIFF unless MEMBER-specific. */
    public static ClusterDiff cluster(Map<String, List<Row>> rowsByMember) {
        List<String> memberIds = new ArrayList<>(rowsByMember.keySet());
        Map<String, Row> first = new LinkedHashMap<>();
        Map<String, Map<String, String>> values = new HashMap<>();
        for (String member : memberIds) {
            for (Row r : rowsByMember.get(member)) {
                first.putIfAbsent(r.key(), r);
                values.computeIfAbsent(r.key(), k -> new HashMap<>()).put(member, r.value());
            }
        }
        int diffCount = 0;
        Set<String> sections = new LinkedHashSet<>();
        for (Map.Entry<String, Row> e : first.entrySet()) {
            if (e.getValue().memberSpecific()) {
                continue;
            }
            Map<String, String> v = values.get(e.getKey());
            boolean diff = v.size() != memberIds.size() || new java.util.HashSet<>(v.values()).size() > 1;
            if (diff) {
                diffCount++;
                sections.add(e.getValue().section());
            }
        }
        return new ClusterDiff(diffCount, first.size(), List.copyOf(sections));
    }
}
